File: packages/planner/src/lookahead.ts
/** Pair-level beam search. Every edge is an executable, preserving F2L insertion. */
import { applyMoves, normalizeOrientation, type CubeState, type Face, type Move } from "@cubing-companion/engine";
import { GEOMETRY, isSlotSolved, slotName, type Slot } from "@cubing-companion/analysis";
import { crossDistance, enumerateF2LInsertion, pairDistance, type SearchResult } from "@cubing-companion/solver";

export interface LookaheadOptions {
  /** Number of additional pairs to plan, including the current pair. */
  readonly depth?: number;
  readonly beamWidth?: number;
  /** Moves above each insertion's optimum to explore. */
  readonly maxExtra?: number;
  /** Total node budget, shared fairly across the first-pair choices. */
  readonly maxNodes?: number;
  readonly maxNodesPerSearch?: number;
}

export interface PairStep {
  /** Slot names and moves in the normalised search frame. */
  readonly slot: string;
  readonly moves: readonly Move[];
  /** Includes any other pairs this insertion happened to solve. */
  readonly solvedSlots: readonly string[];
}

export interface Continuation {
  readonly moves: readonly Move[];
  readonly steps: readonly PairStep[];
  readonly solvedSlots: readonly string[];
}

export interface ContinuationResult {
  /** Complete paths only: a failed search never masquerades as a free continuation. */
  readonly candidates: readonly Continuation[];
  /** Verified shorter prefixes, for inspection plans with a smaller goal. */
  readonly prefixes: readonly Continuation[];
  readonly nodes: number;
  readonly truncated: boolean;
}

export interface PairLookahead {
  readonly slot: Slot;
  /** Null when the budget could not establish a continuation to the requested horizon. */
  readonly plan: Continuation | null;
}

export interface PairLookaheadResult {
  readonly options: readonly PairLookahead[];
  /** Additional pairs each completed plan must solve; reduced near the end of F2L. */
  readonly depth: number;
  readonly nodes: number;
  readonly truncated: boolean;
}

function integer(name: string, value: number, minimum: number, maximum = Infinity): number {
  if (!Number.isSafeInteger(value) || value < minimum || value > maximum) {
    throw new RangeError(`${name} must be an integer between ${minimum} and ${maximum}`);
  }
  return value;
}

interface Path extends Continuation {
  readonly state: CubeState;
}

/**
 * Reach a total number of solved slots. Used for both inspection and pair decisions.
 *
 * Deduplication uses the entire resulting cube, not just the slot: equal-length insertions can
 * leave very different next pairs. A per-depth candidate cap reserves room for +1 alternatives.
 * Beam pruning and node limits make these best-found paths, never a global optimality claim.
 */
export function continueF2L(
  raw: CubeState,
  crossFace: Face,
  targetPairs: number,
  options: LookaheadOptions & { readonly firstSlot?: Slot } = {},
): ContinuationResult {
  integer("targetPairs", targetPairs, 0, 4);
  const beamWidth = integer("beamWidth", options.beamWidth ?? 4, 1);
  const maxExtra = integer("maxExtra", options.maxExtra ?? 1, 0, 3);
  const maxNodes = integer("maxNodes", options.maxNodes ?? 400_000, 0);
  const perSearch = integer("maxNodesPerSearch", options.maxNodesPerSearch ?? 50_000, 1);
  const state = normalizeOrientation(raw);
  if (crossDistance(state, crossFace) !== 0) {
    throw new RangeError("F2L lookahead requires a solved cross");
  }
  const geometry = GEOMETRY[crossFace]!;
  const solved = (position: CubeState) => geometry.slots.filter((slot) => isSlotSolved(position, slot));
  if (options.firstSlot && !geometry.slots.some((s) => slotName(s) === slotName(options.firstSlot!))) {
    throw new RangeError("firstSlot must belong to the selected cross");
  }
  const names = (position: CubeState) => solved(position).map(slotName);
  const lowerBound = (path: Path) => {
    const needed = targetPairs - path.solvedSlots.length;
    if (needed <= 0) return 0;
    const distances = geometry.slots
      .filter((slot) => !isSlotSolved(path.state, slot))
      .map((slot) => pairDistance(path.state, slot))
      .sort((a, b) => a - b);
    // At least this many turns are needed even if other pieces may be disturbed freely.
    return distances[needed - 1]!;
  };
  const accessible = (path: Path) => geometry.slots.reduce((total, slot) => {
    if (isSlotSolved(path.state, slot)) return total;
    return total + Number(geometry.llCorners.includes(path.state.cp.indexOf(slot.corner))) +
      Number(geometry.llEdges.includes(path.state.ep.indexOf(slot.edge)));
  }, 0);
  const compare = (a: Path, b: Path) =>
    a.moves.length + lowerBound(a) - b.moves.length - lowerBound(b) ||
    b.solvedSlots.length - a.solvedSlots.length || accessible(b) - accessible(a);

  let nodes = 0;
  let truncated = false;
  const cache = new Map<string, SearchResult>();
  let beam: Path[] = [{ state, moves: [], steps: [], solvedSlots: names(state) }];
  const completed: Path[] = [];
  const prefixes = new Map<string, Path>();

  // Every insertion solves at least one previously open slot, so four layers suffice.
  for (let level = 0; level <= 4; level++) {
    const next = new Map<string, Path>();
    for (const path of beam) {
      if (path.solvedSlots.length >= targetPairs) {
        completed.push(path);
        continue;
      }
      const open = geometry.slots.filter((slot) => !isSlotSolved(path.state, slot) &&
        (level !== 0 || !options.firstSlot || slotName(slot) === slotName(options.firstSlot)));
      for (const slot of open) {
        const key = `${path.state.key()}:${slotName(slot)}`;
        let result = cache.get(key);
        if (!result) {
          if (nodes >= maxNodes) {
            truncated = true;
            continue;
          }
          result = enumerateF2LInsertion(path.state, crossFace, slot, {
            maxExtra,
            maxSolutions: beamWidth * (maxExtra + 1),
            maxSolutionsPerDepth: beamWidth,
            maxNodes: Math.min(perSearch, maxNodes - nodes),
          });
          nodes += result.stats.nodes;
          truncated ||= result.stats.truncated;
          cache.set(key, result);
        }
        for (const candidate of result.candidates) {
          const after = applyMoves(path.state, candidate.moves);
          const solvedSlots = names(after);
          const child: Path = {
            state: after,
            moves: [...path.moves, ...candidate.moves],
            steps: [...path.steps, { slot: slotName(slot), moves: candidate.moves, solvedSlots }],
            solvedSlots,
          };
          const previous = next.get(after.key());
          if (!previous || compare(child, previous) < 0) next.set(after.key(), child);
          if (solvedSlots.length < targetPairs) {
            const prefix = prefixes.get(after.key());
            if (!prefix || child.moves.length < prefix.moves.length) prefixes.set(after.key(), child);
          }
        }
      }
    }
    const pending: Path[] = [];
    for (const path of next.values()) {
      if (path.solvedSlots.length >= targetPairs) completed.push(path);
      else pending.push(path);
    }
    pending.sort(compare);
    if (pending.length > beamWidth) truncated = true;
    beam = pending.slice(0, beamWidth);
    if (beam.length === 0) break;
  }
  const unique = new Map<string, Path>();
  for (const path of completed.sort(compare)) {
    if (!unique.has(path.state.key())) unique.set(path.state.key(), path);
  }
  return {
    candidates: [...unique.values()].slice(0, beamWidth).map(({ state: _state, ...path }) => path),
    prefixes: [...prefixes.values()].map(({ state: _state, ...path }) => path),
    nodes,
    truncated: truncated || unique.size > beamWidth,
  };
}

/** Compare each first pair over the same horizon, with a separate beam and equal work budget. */
export function lookaheadPairs(
  raw: CubeState,
  crossFace: Face,
  options: LookaheadOptions = {},
): PairLookaheadResult {
  const requestedDepth = integer("depth", options.depth ?? 2, 1, 4);
  const maxNodes = integer("maxNodes", options.maxNodes ?? 1_600_000, 0);
  const state = normalizeOrientation(raw);
  if (crossDistance(state, crossFace) !== 0) throw new RangeError("F2L lookahead requires a solved cross");
  const open = GEOMETRY[crossFace]!.slots.filter((slot) => !isSlotSolved(state, slot));
  const depth = Math.min(requestedDepth, open.length);
  let nodes = 0;
  let truncated = false;
  const results = open.map((slot) => {
    const result = continueF2L(state, crossFace, 4 - open.length + depth, {
      ...options,
      firstSlot: slot,
      maxNodes: Math.floor(maxNodes / open.length),
    });
    nodes += result.nodes;
    truncated ||= result.truncated;
    return { slot, plan: result.candidates[0] ?? null };
  });
  results.sort((a, b) => (a.plan?.moves.length ?? Infinity) - (b.plan?.moves.length ?? Infinity));
  return { options: results, depth, nodes, truncated };
}

File: swift/CubingCore/Sources/CubingCore/Lookahead.swift
/// Bounded beam search through preserving F2L insertions.
public enum Lookahead {
    public struct Options: Sendable {
        public var depth = 2
        public var beamWidth = 4
        public var maxExtra = 1
        public var maxNodes = 400_000
        public var maxNodesPerSearch = 50_000
        public var firstSlot: Slot?
        public init() {}
    }

    public struct PairStep: Sendable {
        public let slot: String
        public let moves: [Move]
        public let solvedSlots: [String]
    }

    public struct Continuation: Sendable {
        public let moves: [Move]
        public let steps: [PairStep]
        public let solvedSlots: [String]
    }

    public struct Result: Sendable {
        public let candidates: [Continuation]
        public let prefixes: [Continuation]
        public let nodes: Int
        public let truncated: Bool
    }

    public struct PairOption: Sendable {
        public let slot: Slot
        public let plan: Continuation?
    }

    public struct PairResult: Sendable {
        public let options: [PairOption]
        public let depth: Int
        public let nodes: Int
        public let truncated: Bool
    }

    private struct Path {
        let state: CubeState
        let moves: [Move]
        let steps: [PairStep]
        let solvedSlots: [String]
        var continuation: Continuation {
            Continuation(moves: moves, steps: steps, solvedSlots: solvedSlots)
        }
    }

    private struct CacheKey: Hashable {
        let state: CubeState
        let slot: String
    }

    public static func continueF2L(
        _ raw: CubeState, _ crossFace: Int, targetPairs: Int, options: Options = Options()
    ) -> Result {
        precondition((0...4).contains(targetPairs))
        precondition(options.beamWidth >= 1 && (0...3).contains(options.maxExtra))
        let state = raw.normalized
        precondition(crossDistance(state, crossFace) == 0, "F2L lookahead requires a solved cross")
        let slots = geometry[crossFace].slots
        if let first = options.firstSlot { precondition(slots.contains(first)) }

        func solved(_ state: CubeState) -> [Slot] { slots.filter { isSlotSolved(state, $0) } }
        func names(_ state: CubeState) -> [String] { solved(state).map(\.name) }
        func lowerBound(_ path: Path) -> Int {
            let needed = targetPairs - path.solvedSlots.count
            if needed <= 0 { return 0 }
            let distances = slots.filter { !isSlotSolved(path.state, $0) }
                .map { pairDistance(path.state, $0) }.sorted()
            return distances[needed - 1]
        }
        func accessible(_ path: Path) -> Int {
            slots.reduce(0) { count, slot in
                if isSlotSolved(path.state, slot) { return count }
                let corner = (0..<8).first { Int(path.state.cp[$0]) == slot.corner }!
                let edge = (0..<12).first { Int(path.state.ep[$0]) == slot.edge }!
                return count + (geometry[crossFace].llCorners.contains(corner) ? 1 : 0)
                    + (geometry[crossFace].llEdges.contains(edge) ? 1 : 0)
            }
        }
        func compare(_ a: Path, _ b: Path) -> Int {
            let score = a.moves.count + lowerBound(a) - b.moves.count - lowerBound(b)
            if score != 0 { return score }
            let solved = b.solvedSlots.count - a.solvedSlots.count
            if solved != 0 { return solved }
            return accessible(b) - accessible(a)
        }
        func stableSort(_ paths: [Path]) -> [Path] {
            paths.enumerated().sorted { a, b in
                let order = compare(a.element, b.element)
                return order == 0 ? a.offset < b.offset : order < 0
            }.map(\.element)
        }

        var nodes = 0, truncated = false
        var cache: [CacheKey: SearchResult] = [:]
        var beam = [Path(state: state, moves: [], steps: [], solvedSlots: names(state))]
        var completed: [Path] = []
        var prefixes: [CubeState: Path] = [:]
        var prefixOrder: [CubeState] = []

        for level in 0...4 {
            var next: [CubeState: Path] = [:]
            var nextOrder: [CubeState] = []
            for path in beam {
                if path.solvedSlots.count >= targetPairs {
                    completed.append(path)
                    continue
                }
                let open = slots.filter {
                    !isSlotSolved(path.state, $0)
                        && (level != 0 || options.firstSlot == nil || $0 == options.firstSlot!)
                }
                for slot in open {
                    let key = CacheKey(state: path.state, slot: slot.name)
                    var result = cache[key]
                    if result == nil {
                        if nodes >= options.maxNodes { truncated = true; continue }
                        var search = insertionDefaults()
                        search.maxExtra = options.maxExtra
                        search.maxSolutions = options.beamWidth * (options.maxExtra + 1)
                        search.maxSolutionsPerDepth = options.beamWidth
                        search.maxNodes = min(options.maxNodesPerSearch, options.maxNodes - nodes)
                        result = enumerateF2LInsertion(path.state, crossFace, slot, options: search)
                        nodes += result!.nodes
                        truncated = truncated || result!.truncated
                        cache[key] = result
                    }
                    for candidate in result!.candidates {
                        let after = path.state.applying(candidate.moves)
                        let solvedSlots = names(after)
                        let child = Path(
                            state: after, moves: path.moves + candidate.moves,
                            steps: path.steps + [PairStep(
                                slot: slot.name, moves: candidate.moves, solvedSlots: solvedSlots)],
                            solvedSlots: solvedSlots)
                        if let previous = next[after] {
                            if compare(child, previous) < 0 { next[after] = child }
                        } else {
                            next[after] = child
                            nextOrder.append(after)
                        }
                        if solvedSlots.count < targetPairs {
                            if let previous = prefixes[after] {
                                if child.moves.count < previous.moves.count { prefixes[after] = child }
                            } else {
                                prefixes[after] = child
                                prefixOrder.append(after)
                            }
                        }
                    }
                }
            }
            var pending: [Path] = []
            for key in nextOrder {
                let path = next[key]!
                if path.solvedSlots.count >= targetPairs { completed.append(path) }
                else { pending.append(path) }
            }
            pending = stableSort(pending)
            if pending.count > options.beamWidth { truncated = true }
            beam = Array(pending.prefix(options.beamWidth))
            if beam.isEmpty { break }
        }
        var unique: [CubeState: Path] = [:]
        var uniqueOrder: [CubeState] = []
        for path in stableSort(completed) where unique[path.state] == nil {
            unique[path.state] = path
            uniqueOrder.append(path.state)
        }
        return Result(
            candidates: uniqueOrder.prefix(options.beamWidth).map { unique[$0]!.continuation },
            prefixes: prefixOrder.map { prefixes[$0]!.continuation },
            nodes: nodes, truncated: truncated || uniqueOrder.count > options.beamWidth)
    }

    /// `lookaheadPairs` in `lookahead.ts`: the whole budget across every open slot.
    ///
    /// `totalNodes` is that budget, split evenly between the slots, and it is separate from
    /// `options.maxNodes` because the TypeScript's defaults differ: 1,600,000 here, 400,000 for a
    /// single `continueF2L`. This once divided `options.maxNodes` — so each slot searched a quarter
    /// of what the TypeScript searches, found different plans wherever the search is budget-bound,
    /// and made the S2 benchmark compare a quarter of the work against the whole of it. The review
    /// oracle (`vectors/review.json`), which records these plans, is what caught it.
    public static let defaultTotalNodes = 1_600_000

    public static func pairs(
        _ raw: CubeState, _ crossFace: Int, options: Options = Options(),
        totalNodes: Int = defaultTotalNodes
    ) -> PairResult {
        let state = raw.normalized
        precondition(crossDistance(state, crossFace) == 0)
        let open = geometry[crossFace].slots.filter { !isSlotSolved(state, $0) }
        let depth = min(options.depth, open.count)
        var nodes = 0, truncated = false
        let ranked = open.enumerated().map { index, slot -> (Int, PairOption) in
            var slotOptions = options
            slotOptions.firstSlot = slot
            slotOptions.maxNodes = totalNodes / open.count
            let result = continueF2L(
                state, crossFace, targetPairs: 4 - open.count + depth, options: slotOptions)
            nodes += result.nodes
            truncated = truncated || result.truncated
            return (index, PairOption(slot: slot, plan: result.candidates.first))
        }.sorted { a, b in
            let left = a.1.plan?.moves.count ?? Int.max
            let right = b.1.plan?.moves.count ?? Int.max
            return left == right ? a.0 < b.0 : left < right
        }
        return PairResult(options: ranked.map(\.1), depth: depth, nodes: nodes, truncated: truncated)
    }
}
